File: 백준/Silver/2583. 영역 구하기/영역 구하기.cc
#include <bits/stdc++.h>

using namespace std;

int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};

int main(int argc, char *argv[]) {
  int n, m, k;
  cin >> n >> m >> k;
  vector<vector<bool>> visited(n, vector<bool>(m, false));

  for (int l = 0; l < k; ++l) {
    int lbx, lby, rtx, rty;
    cin >> lbx >> lby >> rtx >> rty;
    for (int i = lby; i < rty; ++i) {
      for (int j = lbx; j < rtx; ++j) {
        visited[i][j] = true;
      }
    }
  }

  vector<int> ans;
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      if (!visited[i][j]) {
        int size = 1;
        visited[i][j] = true;
        queue<pair<int, int>> q;
        q.push({i, j});

        while (!q.empty()) {
          auto [x, y] = q.front();
          q.pop();

          for (int l = 0; l < 4; ++l) {
            int nx = x + dx[l];
            int ny = y + dy[l];

            if (nx < 0 || ny < 0 || nx >= n || ny >= m || visited[nx][ny])
              continue;
            visited[nx][ny] = true;
            q.push({nx, ny});
            size++;
          }
        }

        ans.emplace_back(size);
      }
    }
  }

  sort(ans.begin(), ans.end());
  cout << ans.size() << "\n";
  for (const auto &elem : ans) {
    cout << elem << " ";
  }
  cout << "\n";

  return 0;
}

File: 백준/Silver/2583. 영역 구하기/영역 구하기.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

var reader *bufio.Reader
var writer *bufio.Writer

var dx []int32 = []int32{-1, 1, 0, 0}
var dy []int32 = []int32{0, 0, -1, 1}

type Pair struct {
	First  int32
	Second int32
}

type Queue struct {
	data []Pair
}

func (q *Queue) Empty() bool {
	return len(q.data) == 0
}

func (q *Queue) Push(p Pair) {
	q.data = append(q.data, p)
}

func (q *Queue) Pop() (Pair, bool) {
	if q.Empty() {
		return Pair{}, false
	}
	front := q.data[0]
	q.data = q.data[1:]
	return front, true
}

func main() {
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, m, k int32
	fmt.Fscan(reader, &n, &m, &k)

	mat := make([][]bool, n)

	for i := int32(0); i < n; i++ {
		mat[i] = make([]bool, m)
	}

	var lbx, lby, rtx, rty int32
	for l := int32(0); l < k; l++ {
		fmt.Fscan(reader, &lbx, &lby, &rtx, &rty)

		for i := lby; i < rty; i++ {
			for j := lbx; j < rtx; j++ {
				mat[i][j] = true
			}
		}
	}

	ans := make([]int32, 0)

	for i := int32(0); i < n; i++ {
		for j := int32(0); j < m; j++ {
			if !mat[i][j] {
				size := int32(1)
				q := Queue{}
				q.Push(Pair{i, j})
				mat[i][j] = true

				for !q.Empty() {
					curr, _ := q.Pop()
					x := curr.First
					y := curr.Second

					for k := 0; k < 4; k++ {
						nx := x + dx[k]
						ny := y + dy[k]

						if nx < 0 || ny < 0 || nx >= n || ny >= m || mat[nx][ny] {
							continue
						}
						mat[nx][ny] = true
						size++
						q.Push(Pair{nx, ny})
					}
				}

				ans = append(ans, size)
			}
		}
	}

	sort.Slice(ans, func(i, j int) bool {
		return ans[i] < ans[j]
	})

	fmt.Fprintln(writer, len(ans))
	for _, v := range ans {
		fmt.Fprintf(writer, "%d ", v)
	}
	fmt.Fprintln(writer)
}